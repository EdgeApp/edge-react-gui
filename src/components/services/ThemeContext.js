// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
