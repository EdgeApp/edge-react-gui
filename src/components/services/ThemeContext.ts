import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
