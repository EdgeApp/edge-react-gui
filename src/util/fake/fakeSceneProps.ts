import { AppParamList, EdgeSceneProps, NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener() {
    return () => undefined
  },
  removeListener() {},
  isFocused() {
    return true
  },

  navigate() {},
  push() {},
  replace() {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  dispatch() {},
  reset() {},
  canGoBack() {
    return true
  },
  getId() {
    return undefined
  },
  getState() {
    return {
      key: 'foo',
      index: 0,
      routes: [],
      routeNames: [],
      type: 'bar',
      stale: false
    }
  },
  getParent() {
    throw new Error('not implemented')
  },
  setOptions() {}
}

export function fakeSceneProps<Name extends keyof AppParamList>(name: Name, params: AppParamList[Name]): EdgeSceneProps<Name> {
  return {
    navigation: fakeNavigation,
    route: {
      name: name,
      key: name + '-0',
      params
    } as any
  }
}
