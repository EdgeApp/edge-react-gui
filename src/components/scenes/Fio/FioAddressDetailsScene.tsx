import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Alert } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { lstrings } from '../../../locales/strings'
import { connect } from '../../../types/reactRedux'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { BUNDLED_TXS_AMOUNT_ALERT, findWalletByFioAddress } from '../../../util/FioAddressUtils'
import { SceneWrapper } from '../../common/SceneWrapper'
import { ConnectWalletsConnector as ConnectWallets } from '../../FioAddress/ConnectWallets'
import { showError } from '../../services/AirshipInstance'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { SettingsHeaderRow } from '../../settings/SettingsHeaderRow'
import { SettingsTappableRow } from '../../settings/SettingsTappableRow'
import { EdgeText } from '../../themed/EdgeText'

interface StateProps {
  fioWallets: EdgeCurrencyWallet[]
}

interface OwnProps extends EdgeSceneProps<'fioAddressDetails'> {}

interface LocalState {
  fioWalletLoading: boolean
  fioWallet: EdgeCurrencyWallet | null
}

type Props = StateProps & ThemeProps & OwnProps

export class FioAddressDetails extends React.Component<Props, LocalState> {
  state: LocalState = {
    fioWalletLoading: false,
    fioWallet: null
  }

  componentDidMount() {
    const { route } = this.props
    const { fioAddressName } = route.params
    if (!fioAddressName) {
      Alert.alert(lstrings.fio_address_details_screen_alert_title, lstrings.fio_address_details_screen_alert_message, [
        { text: lstrings.fio_address_details_screen_alert_button }
      ])
    }
    this.findFioWallet().catch(err => showError(err))
  }

  findFioWallet = async () => {
    const { fioWallets, route } = this.props
    const { fioAddressName } = route.params

    this.setState({ fioWalletLoading: true })
    const fioWallet = await findWalletByFioAddress(fioWallets, fioAddressName)
    this.setState({ fioWalletLoading: false, fioWallet })
  }

  _onPressAccountSettings = (): void => {
    const { navigation, route } = this.props
    const { fioAddressName, bundledTxs } = route.params
    const { fioWallet } = this.state
    if (fioWallet) {
      navigation.navigate('fioAddressSettings', {
        fioWallet,
        fioAddressName,
        bundledTxs,
        refreshAfterAddBundledTxs: true
      })
    } else {
      showError(lstrings.fio_wallet_missing_for_fio_address)
    }
  }

  renderAccountSettings = () => {
    const {
      theme,
      route: {
        params: { bundledTxs }
      }
    } = this.props
    const styles = getStyles(theme)

    if (bundledTxs < BUNDLED_TXS_AMOUNT_ALERT) {
      return (
        <SettingsTappableRow onPress={this._onPressAccountSettings}>
          <IonIcon name="ios-warning" color={theme.warningIcon} style={styles.settingsIcon} />
          <EdgeText style={styles.settingsWarning} numberOfLines={4}>
            {!bundledTxs ? lstrings.fio_address_details_no_bundled_txs : lstrings.fio_address_details_bundled_txs_out_soon}
          </EdgeText>
        </SettingsTappableRow>
      )
    }

    return (
      <SettingsTappableRow label={lstrings.fio_address_details_screen_manage_account_settings} onPress={this._onPressAccountSettings}>
        <IonIcon name="ios-settings" color={theme.icon} style={styles.settingsIcon} />
      </SettingsTappableRow>
    )
  }

  render() {
    const { navigation, theme, route } = this.props
    const { fioAddressName, bundledTxs } = route.params
    const styles = getStyles(theme)
    const bundledTxsLabel = `${lstrings.fio_address_details_screen_bundled_txs}: ${bundledTxs}`

    return (
      <SceneWrapper background="theme">
        <EdgeText style={styles.bundledTxs}>{bundledTxsLabel}</EdgeText>
        {this.renderAccountSettings()}
        <SettingsHeaderRow
          icon={<IonIcon name="ios-link" color={theme.primaryText} size={theme.rem(1.5)} />}
          label={lstrings.fio_address_details_connect_to_wallets}
        />
        <ConnectWallets fioAddressName={fioAddressName} fioWallet={this.state.fioWallet} navigation={navigation} disabled={this.state.fioWalletLoading} />
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  text: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  bundledTxs: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    textAlign: 'center',
    marginTop: theme.rem(-0.5),
    paddingBottom: theme.rem(0.75)
  },
  settingsIcon: {
    fontSize: theme.rem(1.5),
    paddingHorizontal: theme.rem(0.5)
  },
  settingsWarning: {
    fontSize: theme.rem(0.75),
    color: theme.warningText,
    paddingLeft: theme.rem(0.5),
    width: '75%'
  }
}))

export const FioAddressDetailsScene = connect<StateProps, {}, OwnProps>(
  state => ({
    fioWallets: state.ui.wallets.fioWallets
  }),
  dispatch => ({})
)(withTheme(FioAddressDetails))
