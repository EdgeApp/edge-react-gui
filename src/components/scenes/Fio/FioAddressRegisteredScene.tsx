import * as React from 'react'
import { Image, View } from 'react-native'

import { formatDate } from '../../../locales/intl'
import { lstrings } from '../../../locales/strings'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { SceneWrapper } from '../../common/SceneWrapper'
import { FormattedText } from '../../legacy/FormattedText/FormattedText.ui'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { MainButton } from '../../themed/MainButton'

interface OwnProps extends EdgeSceneProps<'fioAddressRegisterSuccess'> {}

type Props = OwnProps & ThemeProps

export class FioAddressRegistered extends React.Component<Props> {
  renderExpDate() {
    const { theme, route } = this.props
    const { expiration } = route.params
    if (expiration != null) {
      const styles = getStyles(theme)

      return (
        <FormattedText style={styles.text}>
          {`${lstrings.fio_address_details_screen_expires} `}
          {formatDate(new Date(expiration))}
        </FormattedText>
      )
    }

    return null
  }

  render() {
    const { theme, navigation, route } = this.props
    const { fioName } = route.params
    const styles = getStyles(theme)

    return (
      <SceneWrapper background="theme">
        <View style={styles.view}>
          <View style={styles.texts}>
            <View style={styles.image}>
              <Image source={theme.fioAddressLogo} />
            </View>
            <FormattedText style={styles.text}>{lstrings.fio_address_details_screen_registered}</FormattedText>
            <FormattedText style={styles.title}>{fioName}</FormattedText>
            {this.renderExpDate()}
          </View>
          <MainButton marginRem={2} onPress={() => navigation.navigate('fioAddressList', {})} label={lstrings.title_fio_names} />
        </View>
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  view: {
    flex: 2,
    flexDirection: 'column',
    justifyContent: 'space-between',
    paddingHorizontal: theme.rem(1)
  },
  texts: {
    flex: 1,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: theme.rem(2.5)
  },
  text: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  image: {
    marginLeft: theme.rem(0.25),
    marginBottom: theme.rem(3)
  },
  title: {
    fontSize: theme.rem(1.75),
    color: theme.primaryText,
    marginTop: theme.rem(1.25),
    marginBottom: theme.rem(0.75)
  },
  expiration: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    textAlign: 'center',
    marginTop: theme.rem(-0.5),
    paddingBottom: theme.rem(0.75)
  }
}))

export const FioAddressRegisteredScene = withTheme(FioAddressRegistered)
