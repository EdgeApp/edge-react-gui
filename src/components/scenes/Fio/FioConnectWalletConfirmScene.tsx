import * as React from 'react'
import { ScrollView } from 'react-native'

import { lstrings } from '../../../locales/strings'
import { CcWalletMap } from '../../../reducers/FioReducer'
import { connect } from '../../../types/reactRedux'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { FioConnectionWalletItem } from '../../../types/types'
import { FIO_NO_BUNDLED_ERR_CODE, updatePubAddressesForFioAddress } from '../../../util/FioAddressUtils'
import { SceneWrapper } from '../../common/SceneWrapper'
import { ButtonsModal } from '../../modals/ButtonsModal'
import { Airship, showError, showToast } from '../../services/AirshipInstance'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { EdgeText } from '../../themed/EdgeText'
import { Slider } from '../../themed/Slider'
import { Radio } from '../../themed/ThemedButtons'
import { Tile } from '../../tiles/Tile'

interface State {
  acknowledge: boolean
  connectWalletsLoading: boolean
  showSlider: boolean
}

interface StateProps {
  ccWalletMap: CcWalletMap
  isConnected: boolean
}

interface OwnProps extends EdgeSceneProps<'fioConnectToWalletsConfirm'> {}

interface DispatchProps {
  updateConnectedWallets: (fioAddress: string, ccWalletMap: CcWalletMap) => void
}

type Props = StateProps & DispatchProps & OwnProps & ThemeProps

export class FioConnectWalletConfirm extends React.Component<Props, State> {
  state = {
    acknowledge: false,
    connectWalletsLoading: false,
    showSlider: true
  }

  confirm = async (): Promise<void> => {
    const { updateConnectedWallets, ccWalletMap, isConnected, navigation, route } = this.props
    const { fioWallet, fioAddressName, walletsToConnect, walletsToDisconnect } = route.params
    if (isConnected) {
      this.setState({ connectWalletsLoading: true })
      const newCcWalletMap = { ...ccWalletMap }
      try {
        let promiseArray = walletsToConnect.map(async (wallet: FioConnectionWalletItem) => ({
          walletId: wallet.id,
          tokenCode: wallet.currencyCode,
          chainCode: wallet.chainCode,
          publicAddress: (await wallet.edgeWallet.getReceiveAddress()).publicAddress
        }))

        let publicAddresses = await Promise.all(promiseArray)

        const { updatedCcWallets, error } = await updatePubAddressesForFioAddress(fioWallet, fioAddressName, publicAddresses)
        if (updatedCcWallets.length) {
          for (const { fullCurrencyCode, walletId } of updatedCcWallets) {
            newCcWalletMap[fullCurrencyCode] = walletId
          }
          updateConnectedWallets(fioAddressName, newCcWalletMap)
        }

        promiseArray = walletsToDisconnect.map(async (wallet: FioConnectionWalletItem) => ({
          walletId: wallet.id,
          tokenCode: wallet.currencyCode,
          chainCode: wallet.chainCode,
          publicAddress: (await wallet.edgeWallet.getReceiveAddress()).publicAddress
        }))

        publicAddresses = await Promise.all(promiseArray)

        const { updatedCcWallets: removedCcWallets, error: removedError } = await updatePubAddressesForFioAddress(
          fioWallet,
          fioAddressName,
          publicAddresses,
          false
        )
        if (removedCcWallets.length) {
          for (const { fullCurrencyCode } of removedCcWallets) {
            newCcWalletMap[fullCurrencyCode] = ''
          }
          updateConnectedWallets(fioAddressName, newCcWalletMap)
        }

        const eitherError = error ?? removedError
        if (eitherError != null) {
          const walletsToConnectLeft: FioConnectionWalletItem[] = []
          const walletsToDisconnectLeft: FioConnectionWalletItem[] = []
          if (updatedCcWallets.length) {
            for (const walletToConnect of walletsToConnect) {
              if (
                updatedCcWallets.findIndex(
                  ({ walletId, fullCurrencyCode }) => walletId === walletToConnect.id && fullCurrencyCode === walletToConnect.fullCurrencyCode
                ) < 0
              ) {
                walletsToConnectLeft.push(walletToConnect)
              }
            }
          }
          if (removedCcWallets.length) {
            for (const walletToDisconnect of walletsToDisconnect) {
              if (
                removedCcWallets.findIndex(
                  ({ walletId, fullCurrencyCode }) => walletId === walletToDisconnect.id && fullCurrencyCode === walletToDisconnect.fullCurrencyCode
                ) < 0
              ) {
                walletsToDisconnectLeft.push(walletToDisconnect)
              }
            }
          }
          if (walletsToConnectLeft.length || walletsToDisconnectLeft.length) {
            navigation.setParams({ fioWallet, fioAddressName, walletsToConnect: walletsToConnectLeft, walletsToDisconnect: walletsToDisconnectLeft })
            this.resetSlider()
          }
          throw eitherError
        }
        if (walletsToConnect.length) {
          showToast(lstrings.fio_connect_wallets_success)
        } else {
          if (walletsToDisconnect.length) showToast(lstrings.fio_disconnect_wallets_success)
        }
        navigation.goBack()
      } catch (e: any) {
        if (e.code === FIO_NO_BUNDLED_ERR_CODE) {
          this.setState({ connectWalletsLoading: false })
          const answer = await Airship.show<'ok' | undefined>(bridge => (
            <ButtonsModal
              bridge={bridge}
              title={lstrings.fio_no_bundled_err_msg}
              message={lstrings.fio_no_bundled_add_err_msg}
              buttons={{
                ok: { label: lstrings.title_fio_add_bundled_txs }
              }}
              closeArrow
            />
          ))
          if (answer === 'ok') {
            navigation.navigate('fioAddressSettings', {
              showAddBundledTxs: true,
              fioWallet,
              fioAddressName: fioAddressName
            })
          }
          return
        }
        this.resetSlider()
        showError(e)
      }
      this.setState({ connectWalletsLoading: false })
    } else {
      showError(lstrings.fio_network_alert_text)
    }
  }

  resetSlider = (): void => {
    this.setState({ showSlider: false }, () => this.setState({ showSlider: true }))
  }

  check = (): void => {
    const { acknowledge } = this.state

    this.setState({ acknowledge: !acknowledge })
  }

  renderWalletLine = (wallet: FioConnectionWalletItem) => {
    const styles = getStyles(this.props.theme)
    const label = `${wallet.name} (${wallet.currencyCode})`
    return (
      <EdgeText key={`${wallet.id}-${wallet.currencyCode}`} style={styles.content}>
        {label}
      </EdgeText>
    )
  }

  render() {
    const { theme, route } = this.props
    const { fioAddressName, walletsToConnect, walletsToDisconnect } = route.params
    const { acknowledge, connectWalletsLoading, showSlider } = this.state
    const styles = getStyles(theme)

    return (
      <SceneWrapper background="theme">
        <ScrollView>
          <Tile type="static" title={lstrings.fio_address_register_form_field_label} body={fioAddressName} />
          {walletsToConnect.length ? (
            <Tile type="static" title={lstrings.title_fio_connect_to_wallet}>
              {walletsToConnect.map(this.renderWalletLine)}
            </Tile>
          ) : null}

          {walletsToDisconnect.length ? (
            <Tile type="static" title={lstrings.title_fio_disconnect_wallets}>
              {walletsToDisconnect.map(this.renderWalletLine)}
            </Tile>
          ) : null}

          <Radio value={acknowledge} onPress={this.check} marginRem={[2, 2, 0]}>
            <EdgeText style={styles.checkTitle} numberOfLines={4}>
              {lstrings.fio_connect_checkbox_text}
            </EdgeText>
          </Radio>
          {showSlider && (
            <Slider
              parentStyle={styles.slider}
              onSlidingComplete={this.confirm}
              disabled={!acknowledge || connectWalletsLoading}
              disabledText={lstrings.send_confirmation_slide_to_confirm}
              showSpinner={connectWalletsLoading}
            />
          )}
        </ScrollView>
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  content: {
    color: theme.primaryText,
    fontSize: theme.rem(1),
    marginHorizontal: theme.rem(0.25),
    textAlign: 'left'
  },
  checkTitle: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    marginLeft: theme.rem(1)
  },
  slider: {
    paddingVertical: theme.rem(2)
  }
}))

export const FioConnectWalletConfirmScene = connect<StateProps, DispatchProps, OwnProps>(
  (state, { route: { params } }) => ({
    ccWalletMap: state.ui.fio.connectedWalletsByFioAddress[params.fioAddressName],
    isConnected: state.network.isConnected
  }),
  dispatch => ({
    updateConnectedWallets(fioAddress: string, ccWalletMap: CcWalletMap) {
      dispatch({
        type: 'FIO/UPDATE_CONNECTED_WALLETS_FOR_FIO_ADDRESS',
        data: { fioAddress, ccWalletMap }
      })
    }
  })
)(withTheme(FioConnectWalletConfirm))
