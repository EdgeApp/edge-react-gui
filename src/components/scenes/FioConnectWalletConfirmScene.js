// @flow

import * as React from 'react'
import { ScrollView } from 'react-native'

import s from '../../locales/strings.js'
import { FIO_NO_BUNDLED_ERR_CODE, updatePubAddressesForFioAddress } from '../../modules/FioAddress/util'
import { Slider } from '../../modules/UI/components/Slider/Slider.js'
import type { CcWalletMap } from '../../reducers/FioReducer'
import { connect } from '../../types/reactRedux.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import type { FioConnectionWalletItem } from '../../types/types'
import { SceneWrapper } from '../common/SceneWrapper'
import { ButtonsModal } from '../modals/ButtonsModal'
import { Airship, showError, showToast } from '../services/AirshipInstance'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { EdgeText } from '../themed/EdgeText'
import { Radio } from '../themed/ThemedButtons'
import { Tile } from '../themed/Tile'

type State = {
  acknowledge: boolean,
  connectWalletsLoading: boolean,
  showSlider: boolean
}

type StateProps = {
  ccWalletMap: CcWalletMap,
  isConnected: boolean
}

type OwnProps = {
  navigation: NavigationProp<'fioConnectToWalletsConfirm'>,
  route: RouteProp<'fioConnectToWalletsConfirm'>
}

type DispatchProps = {
  updateConnectedWallets: (fioAddress: string, ccWalletMap: CcWalletMap) => void
}

type Props = StateProps & DispatchProps & OwnProps & ThemeProps

export class FioConnectWalletConfirm extends React.Component<Props, State> {
  state = {
    acknowledge: false,
    connectWalletsLoading: false,
    showSlider: true
  }

  confirm = async (): Promise<void> => {
    const { updateConnectedWallets, ccWalletMap, isConnected, navigation, route } = this.props
    const { fioWallet, fioAddressName, walletsToConnect, walletsToDisconnect } = route.params
    if (isConnected) {
      this.setState({ connectWalletsLoading: true })
      const newCcWalletMap = { ...ccWalletMap }
      try {
        const { updatedCcWallets, error } = await updatePubAddressesForFioAddress(
          fioWallet,
          fioAddressName,
          walletsToConnect.map((wallet: FioConnectionWalletItem) => ({
            walletId: wallet.id,
            tokenCode: wallet.currencyCode,
            chainCode: wallet.chainCode,
            publicAddress: wallet.publicAddress
          }))
        )
        if (updatedCcWallets.length) {
          for (const { fullCurrencyCode, walletId } of updatedCcWallets) {
            newCcWalletMap[fullCurrencyCode] = walletId
          }
          updateConnectedWallets(fioAddressName, newCcWalletMap)
        }

        const { updatedCcWallets: removedCcWallets, error: removedError } = await updatePubAddressesForFioAddress(
          fioWallet,
          fioAddressName,
          walletsToDisconnect.map((wallet: FioConnectionWalletItem) => ({
            walletId: wallet.id,
            tokenCode: wallet.currencyCode,
            chainCode: wallet.chainCode,
            publicAddress: wallet.publicAddress
          })),
          false
        )
        if (removedCcWallets.length) {
          for (const { fullCurrencyCode } of removedCcWallets) {
            newCcWalletMap[fullCurrencyCode] = ''
          }
          updateConnectedWallets(fioAddressName, newCcWalletMap)
        }

        if (error || removedError) {
          const walletsToConnectLeft = []
          const walletsToDisconnectLeft = []
          if (updatedCcWallets.length) {
            for (const walletToConnect of walletsToConnect) {
              if (
                updatedCcWallets.findIndex(
                  ({ walletId, fullCurrencyCode }) => walletId === walletToConnect.id && fullCurrencyCode === walletToConnect.fullCurrencyCode
                ) < 0
              ) {
                walletsToConnectLeft.push(walletToConnect)
              }
            }
          }
          if (removedCcWallets.length) {
            for (const walletToDisconnect of walletsToDisconnect) {
              if (
                removedCcWallets.findIndex(
                  ({ walletId, fullCurrencyCode }) => walletId === walletToDisconnect.id && fullCurrencyCode === walletToDisconnect.fullCurrencyCode
                ) < 0
              ) {
                walletsToDisconnectLeft.push(walletToDisconnect)
              }
            }
          }
          if (walletsToConnectLeft.length || walletsToDisconnectLeft.length) {
            navigation.setParams({ fioWallet, fioAddressName, walletsToConnect: walletsToConnectLeft, walletsToDisconnect: walletsToDisconnectLeft })
            this.resetSlider()
          }
          throw error || removedError
        }
        if (walletsToConnect.length) {
          showToast(s.strings.fio_connect_wallets_success)
        } else {
          if (walletsToDisconnect.length) showToast(s.strings.fio_disconnect_wallets_success)
        }
        navigation.goBack()
      } catch (e) {
        if (e.code === FIO_NO_BUNDLED_ERR_CODE) {
          this.setState({ connectWalletsLoading: false })
          const answer = await Airship.show(bridge => (
            <ButtonsModal
              bridge={bridge}
              title={s.strings.fio_no_bundled_err_msg}
              message={s.strings.fio_no_bundled_add_err_msg}
              buttons={{
                ok: { label: s.strings.title_fio_add_bundled_txs }
              }}
              closeArrow
            />
          ))
          if (answer === 'ok') {
            navigation.navigate('fioAddressSettings', {
              showAddBundledTxs: true,
              fioWallet,
              fioAddressName: fioAddressName
            })
          }
          return
        }
        this.resetSlider()
        showError(e)
      }
      this.setState({ connectWalletsLoading: false })
    } else {
      showError(s.strings.fio_network_alert_text)
    }
  }

  resetSlider = (): void => {
    this.setState({ showSlider: false }, () => this.setState({ showSlider: true }))
  }

  check = (): void => {
    const { acknowledge } = this.state

    this.setState({ acknowledge: !acknowledge })
  }

  renderWalletLine = (wallet: FioConnectionWalletItem) => {
    const styles = getStyles(this.props.theme)
    const label = `${wallet.name} (${wallet.currencyCode})`
    return (
      <EdgeText key={`${wallet.id}-${wallet.currencyCode}`} style={styles.content}>
        {label}
      </EdgeText>
    )
  }

  render() {
    const { theme, route } = this.props
    const { fioAddressName, walletsToConnect, walletsToDisconnect } = route.params
    const { acknowledge, connectWalletsLoading, showSlider } = this.state
    const styles = getStyles(theme)

    return (
      <SceneWrapper background="header">
        <ScrollView>
          <Tile type="static" title={s.strings.fio_address_register_form_field_label} body={fioAddressName} />
          {walletsToConnect.length ? (
            <Tile type="static" title={s.strings.title_fio_connect_to_wallet}>
              {walletsToConnect.map(this.renderWalletLine)}
            </Tile>
          ) : null}

          {walletsToDisconnect.length ? (
            <Tile type="static" title={s.strings.title_fio_disconnect_wallets}>
              {walletsToDisconnect.map(this.renderWalletLine)}
            </Tile>
          ) : null}

          <Radio value={acknowledge} onPress={this.check} marginRem={[2, 2, 0]}>
            <EdgeText style={styles.checkTitle} numberOfLines={4}>
              {s.strings.fio_connect_checkbox_text}
            </EdgeText>
          </Radio>
          {showSlider && (
            <Slider
              parentStyle={styles.slider}
              onSlidingComplete={this.confirm}
              disabled={!acknowledge || connectWalletsLoading}
              disabledText={s.strings.send_confirmation_slide_to_confirm}
              showSpinner={connectWalletsLoading}
            />
          )}
        </ScrollView>
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  content: {
    color: theme.primaryText,
    fontSize: theme.rem(1),
    marginHorizontal: theme.rem(0.25),
    textAlign: 'left'
  },
  checkTitle: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    marginLeft: theme.rem(1)
  },
  slider: {
    paddingVertical: theme.rem(2)
  }
}))

export const FioConnectWalletConfirmScene = connect<StateProps, DispatchProps, OwnProps>(
  (state, { route: { params } }) => ({
    ccWalletMap: state.ui.fio.connectedWalletsByFioAddress[params.fioAddressName],
    isConnected: state.network.isConnected
  }),
  dispatch => ({
    updateConnectedWallets(fioAddress: string, ccWalletMap: CcWalletMap) {
      dispatch({
        type: 'FIO/UPDATE_CONNECTED_WALLETS_FOR_FIO_ADDRESS',
        data: { fioAddress, ccWalletMap }
      })
    }
  })
)(withTheme(FioConnectWalletConfirm))
