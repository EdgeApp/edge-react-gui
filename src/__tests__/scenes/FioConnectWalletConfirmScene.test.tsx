import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { FioConnectWalletConfirm } from '../../components/scenes/Fio/FioConnectWalletConfirmScene'
import { getTheme } from '../../components/services/ThemeContext'
import { fakeSceneProps } from '../../util/fake/fakeSceneProps'

describe('FioConnectWalletConfirm', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const fakeWallet: any = {
      currencyCode: 'FIO',
      nativeAmount: '100',
      networkFee: '1',
      blockHeight: 34,
      date: 220322,
      txid: '0x34346463',
      signedTx: '0xdgs3442',
      ourReceiveAddresses: ['FioAddress']
    }

    const actual = renderer.render(
      <FioConnectWalletConfirm
        {...fakeSceneProps('fioConnectToWalletsConfirm', {
          fioWallet: fakeWallet,
          fioAddressName: 'MyFioAddress',
          walletsToConnect: [],
          walletsToDisconnect: []
        })}
        ccWalletMap={['FIO'] as any}
        isConnected
        updateConnectedWallets={(fioAddress, ccWalletMap) => undefined}
        theme={getTheme()}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
