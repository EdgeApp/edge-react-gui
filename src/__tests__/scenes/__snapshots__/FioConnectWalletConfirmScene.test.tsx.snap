// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FioConnectWalletConfirm should render with loading props 1`] = `
<SceneWrapper
  background="header"
>
  <ScrollView>
    <WithTheme(TileComponent)
      body="MyFioAddress"
      title="FIO Crypto Handle"
      type="static"
    />
    <Radio
      marginRem={
        Array [
          2,
          2,
          0,
        ]
      }
      onPress={[Function]}
      value={false}
    >
      <WithTheme(EdgeTextComponent)
        numberOfLines={4}
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 17,
            "marginLeft": 22,
          }
        }
      >
        I acknowledge that one address from each wallet will become public. These addresses will remain public even after wallets are disconnected.
      </WithTheme(EdgeTextComponent)>
    </Radio>
    <WithTheme(SliderComponent)
      disabled={true}
      disabledText="Slide to Confirm"
      onSlidingComplete={[Function]}
      parentStyle={
        Object {
          "paddingVertical": 45,
        }
      }
      showSpinner={false}
    />
  </ScrollView>
</SceneWrapper>
`;
